core/executor/src/wasm_runtimes_cache.rs

```diff
@@ -0,0 +1,316 @@
+use primitives::Blake2Hasher;
+use runtime_version::RuntimeVersion;
+use state_machine::Externalities;
+use std::collections::hash_map::{HashMap, Entry};
+use std::mem;
+use wasmi::{Module as WasmModule, ModuleRef as WasmModuleInstanceRef, RuntimeValue};
+
+#[derive(Debug)]
+enum CacheError {
+	CodeNotFound,
+	ApplySnapshotFailed,
+	InvalidModule,
+	CantDeserializeWasm,
+	Instantiation,
+}
+
+/// A runtime along with its version and initial state snapshot.
+#[derive(Clone)]
+struct CachedRuntime {
+	/// A wasm module instance.
+	instance: WasmModuleInstanceRef,
+	/// Runtime version according to `Core_version`.
+	///
+	/// Can be `None` if the runtime doesn't expose this function.
+	version: Option<RuntimeVersion>,
+	/// The snapshot of the instance's state taken just after the instantiation.
+	state_snapshot: StateSnapshot,
+}
+
+/// A state snapshot of an instance taken just after instantiation.
+///
+/// It is used for restoring the state of the module after execution.
+#[derive(Clone)]
+struct StateSnapshot {
+	/// The offset and the contents of the memory segments that should be copied at
+	/// to restore the snapshot.
+	data_segments: Vec<(u32, Vec<u8>)>,
+	/// The list of all global variables of the module in their sequential order.
+	global_mut_values: Vec<RuntimeValue>,
+	heap_pages: u32,
+}
+
+impl StateSnapshot {
+	// Returns `None` if instance is not valid.
+	fn take(
+		module_instance: &WasmModuleInstanceRef,
+		data_segments: Vec<DataSegment>,
+		heap_pages: u32,
+	) -> Option<Self> {
+		let mut prepared_segments = Vec::with_capacity(data_segments.len());
+		for mut segment in data_segments {
+			// Just replace contents of the segment since the segments will be discarded later
+			// anyway.
+			let contents = mem::replace(segment.value_mut(), vec![]);
+
+			let init_expr = segment.offset().code();
+			// [op, End]
+			if init_expr.len() != 2 {
+				return None;
+			}
+			let offset = match init_expr[0] {
+				Instruction::I32Const(v) => v as u32,
+				Instruction::GetGlobal(idx) => {
+					let global_val = module_instance.globals().get(idx as usize)?.get();
+					match global_val {
+						RuntimeValue::I32(v) => v as u32,
+						_ => return None,
+					}
+				}
+				_ => return None,
+			};
+			prepared_segments.push((offset, contents))
+		}
+
+		// Collect all values of mutable globals.
+		let global_mut_values = module_instance
+			.globals()
+			.iter()
+			.filter(|g| g.is_mutable())
+			.map(|g| g.get())
+			.collect();
+
+		Some(Self {
+			data_segments: prepared_segments,
+			global_mut_values,
+			heap_pages,
+		})
+	}
+
+	/// Reset the runtime instance to the initial version by restoring
+	/// the preserved memory and globals.
+	///
+	/// Returns `Err` if applying the snapshot is failed.
+	fn apply(&self, instance: &WasmModuleInstanceRef) -> Result<(), CacheError> {
+		let memory = instance
+			.export_by_name("memory")
+			.ok_or(CacheError::ApplySnapshotFailed)?
+			.as_memory()
+			.cloned()
```

Comment:
COuld use `and_then` as well to have just one `ok_or`.

Reply:
Haha, tried to eradicate `and_then`'s since I have (small) allergy on them. Put a thumbs up on this comment if you feel strongly about this.